.github/workflows/commits.yml

```
@@ -0,0 +1,64 @@
+# Checks all commits in a PR follow the repo rules:
+#
+#   1. conventional commit messages
+#   2. used `git commit --signoff`
+#   3. no extra merge commits
+name: Commits
+
+on:
+  workflow_dispatch:
+  # Perform these checks on PRs into *any* branch.
+  #
+  # Motivation:
+  #   Commits which are not --signoff but merged into other branches
+  #   will likely make their way into PRs for master. At which
+  #   point it will be difficult to get the original author to --signoff.
+  pull_request:
+
+jobs:
+  commit-checks:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+
+      - name: Install conventional-commit linter
+        run: |
+          # Intall the linter and the conventional commits config
+          npm install @commitlint/config-conventional @commitlint/cli
+
+          # Extend the conventional commits config with the `--signoff`
+          # requirement.
+          echo "module.exports = {
+            extends: ['@commitlint/config-conventional'],
+            rules: {
+              'signed-off-by': [2, 'always', 'Signed-off-by:'],
+            }
+          }" > commitlint.config.js
+
+      - name: Conventional commit check
+        run: |
+          # At this stage git's HEAD does not refer to the latest commit in the PR,
+          # but rather to the merge commit of the PR. So instead we have to get 'HEAD'
+          # from the PR event.
+          PR_HEAD=${{ github.event.pull_request.head.sha }}
+          npx commitlint --to $PR_HEAD --from $PR_HEAD~${{ github.event.pull_request.commits }}
+
+      - name: No merge commits
+        run: |
+          # At this stage git's HEAD does not refer to the latest commit in the PR,
+          # but rather to the merge commit of the PR. So instead we have to get 'HEAD'
+          # from the PR event.
+          PR_HEAD=${{ github.event.pull_request.head.sha }}
+
+          # This will list any merge commits in the PR commit path
+          MERGE=$(git log --merges --ancestry-path $PR_HEAD~${{ github.event.pull_request.commits }}..$PR_HEAD)
+
+          # The merge list should be empty
+          [[ ! -z "$MERGE" ]] && {
+            echo "PR contains merge commits:";
+            echo $MERGE;
+            exit 1;
+          }
+          exit 0;
```

```suggestion
          exit 0;

```

Not sure if my suggestion works; github and git complain whenever text files don't end with newline which makes them binary files according to some.

---

That's.. interesting.

---

Recommend you to find this option in your `$EDITOR` for future. In my global editorconfig it's just:

```
[*]
insert_final_newline = true
```

But I think this is a noop with vim/neovim.